Detectors/CTP/reconstruction/include/CTPReconstruction/CTFCoder.h

```diff
@@ -151,27 +163,118 @@ o2::ctf::CTFIOSize CTFCoder::decode(const CTF::base& ec, VTRG& data, LumiInfo& l
     } else {
       ir.bc += bcInc[itrig];
     }
-    if (checkIROK || canApplyBCShift(ir)) { // correction will be ok
-      checkIROK = true;
+    if (checkIROKInputs || canApplyBCShiftInputs(ir)) { // correction will be ok
+      // checkIROK = true;
+      auto irs = ir - mBCShiftInputs;
+      uint64_t CTPInputMask = 0;
+      for (int i = 0; i < CTFHelper::CTPInpNBytes; i++) {
+        CTPInputMask |= static_cast<uint64_t>(*itInp++) << (8 * i);
+      }
+      if (CTPInputMask) {
+        if (digitsMap.count(irs)) {
+          if (digitsMap[irs].isInputEmpty()) {
+            digitsMap[irs].CTPInputMask = CTPInputMask;
+            // LOG(info) << "IR1:";
+            // digitsMap[irs].printStream(std::cout);
+          } else {
+            LOG(error) << "CTPInpurMask already exist:" << irs << " dig.CTPInputMask:" << digitsMap[irs].CTPInputMask << " CTPInputMask:" << CTPInputMask;
+          }
+        } else {
+          CTPDigit dig = {irs, CTPInputMask, 0};
+          digitsMap[irs] = dig;
+          // LOG(info) << "IR2:";
+          // digitsMap[irs].printStream(std::cout);
+        }
+      }
     } else { // correction would make IR prior to mFirstTFOrbit, skip
       itInp += CTFHelper::CTPInpNBytes;
+      // itCls += CTFHelper::CTPClsNBytes;
+      // continue;
+    }
+    if (checkIROK || canApplyBCShift(ir)) { // correction will be ok
+      // checkIROK = true;
+      auto irs = ir - mBCShift;
+      uint64_t CTPClassMask = 0;
+      for (int i = 0; i < CTFHelper::CTPClsNBytes; i++) {
+        CTPClassMask |= static_cast<uint64_t>(*itCls++) << (8 * i);
+      }
+      if (CTPClassMask) {
+        if (digitsMap.count(irs)) {
+          if (digitsMap[irs].isClassEmty()) {
+            digitsMap[irs].CTPClassMask = CTPClassMask;
+            // LOG(info) << "TCM1:";
+            // digitsMap[irs].printStream(std::cout);
+          } else {
+            LOG(error) << "CTPClassMask already exist:" << irs << " dig.CTPClassMask:" << digitsMap[irs].CTPClassMask << " CTPClassMask:" << CTPClassMask;
+          }
+        } else {
+          CTPDigit dig = {irs, 0, CTPClassMask};
+          digitsMap[irs] = dig;
+          // LOG(info) << "TCM2:";
+          // digitsMap[irs].printStream(std::cout);
+        }
+      }
+    } else { // correction would make IR prior to mFirstTFOrbit, skip
+      // itInp += CTFHelper::CTPInpNBytes;
       itCls += CTFHelper::CTPClsNBytes;
-      continue;
+      // continue;
     }
-    auto& dig = data.emplace_back();
-    dig.intRecord = ir - mBCShift;
-    for (int i = 0; i < CTFHelper::CTPInpNBytes; i++) {
-      dig.CTPInputMask |= static_cast<uint64_t>(*itInp++) << (8 * i);
+  }
+  if (mDecodeInps) {
+    std::vector<CTPDigit> digits;
+    o2::ctp::RawDataDecoder::shiftInputs(digitsMap, digits, mFirstTFOrbit);
```

Comment:
in principle, one can avoid creating a local digits vector and then copying it to VTRG& data: better to template the https://github.com/AliceO2Group/AliceO2/blob/fcac2f96ff8a7dd9ae031d6a83c5d84faf463773/Detectors/CTP/reconstruction/src/RawDataDecoder.cxx#L348C1-L348C1 with `VTRG& digits` instead of `std::vector<CTPDigit>& digits` argument (need to move it to the header) and fill the final vector directly. 

Or, one can avoid templating by simply using `o2::pmr::vector<..>` instead of `std::vector<..>` and fill it directly also in the RawDecoderSpec.

If in doubt, I can do this later.

Reply:
I tried to do o2::pmr - just replaced everywhere ....